app.py

```diff
@@ -459,6 +467,20 @@ def delete_comment(id):
         flash("You are not authorized to delete this comment", "danger")
     return redirect(url_for("post", id=comment.post_id))
 
+@app.route("/posts/<id>/flag", methods=['POST', 'GET'])
+def report(id):
+    post = Post.query.filter_by(id=id).first()
+    ref = request.args.get('ref')
+    if post:
+        existing_flags = Flag.query.filter_by(posts_id=id, author_id=current_user.id).first()
```

Comment:
you can always define additional methods on the model, here. for full points, I would think of where this logic belongs. my intuition is that it belongs to the `User` class, so on the `User` model, I would define a new method:

```
class User(db.Model):
  ## blah blah

  def has_flagged(self, post):
    return Flag.query.filter_by(post_id=post.id, author_id=self.id).first() is not None
```

why is it `posts_id` and not `post_id`, by the way?

then if you have this method, this line looks like:

```
if post is not None:
   if current_user.has_flagged(post):
       flash("error you've already flagged this post")
       return redirect(ref)
        # alternatively, you can throw an error here like 422 bad request
    else:
        # the code you have here, it's great, except maybe 
         flag = Flag(posts_id=post.id, user_id=current_user.id)
```